packages/@lwc/engine-core/src/framework/utils.ts

```diff
@@ -85,3 +87,19 @@ export function cloneAndOmitKey(object: { [key: string]: any }, keyToOmit: strin
     }
     return result;
 }
+
+// Set a ref (lwc:ref) on a VM, from a template API
+export function setRef(vm: VM, ref: string | undefined, vnode: VNode) {
+    if (!isUndefined(ref)) {
+        if (isNull(vm.refs)) {
+            // Create the refs object on-demand. This ensures that templates with no refs
+            // don't pay the perf tax of creating objects unnecessarily.
+            vm.refs = Object.create(null);
```

Comment:
All the template refs are known in advance. Could the template compiler attach a list of ref names (if any) on the template function directly? The ref object could then be created when the template is evaluated.

Reply:
On second thought, I think the current approach may be preferable. For cases like `<template if:true>`/`<template if:false>`, we actually don't know until runtime which refs will be defined. (One ref could be in the `true` block, another one could be in the `false` block.) I think building up the `refs` object at runtime may be the cleanest.

Reply:
Related to your comment: https://github.com/salesforce/lwc-rfcs/pull/59/files#r770847293

> > Shall all the know refs in the template be set to `undefined` in advance?
> 
> Yes. A common pattern developers will use is something like `if (this.refs.foo !== undefined) { /* ... */ }`, to check if the node exists yet or not. We already see this pattern in the wild with `querySelector` and `firstChild`.

The current approach contradicts the comment you here. The `Object.keys(this.refs)` would only contain the refs currently rendered and not all the potential refs from the template.